### ISSUE TITLE
External db changes to document nested objects aren't sent over subscription to client

### ISSUE THREAD

**Issue by igal1c0de4n** [meteor/meteor#12119]
This issue can be reproduced on Meteor ver 2.7.3 on macos, linux (aws ec2) + mongodb v5.0.7 and 5.0.10.
The first Meteor version having this issue is unknown, partially because I started seeing this issue after upgrading from mongodb v4.4.

### How to reproduce
- git clone the Tasks sample app https://github.com/fredmaiaarantes/simpletasks 
- Run against mongodb v5.0.7 or newer (it cannot be reproduced on v4.4)
Note: be sure to provide a valid `MONGO_OPLOG_URL` environment variable.
Be advised oplog url strings may contain special characters such as & and ?
Those characters break the string if provided in the command line without quotes. 
Hence confirm this env var is quoted:
```
MONGO_OPLOG_URL="mongodb://user:psw@localhost:27017/local?replicaSet=001-rs&authSource=main"
```
If there's any problem with the oplog string, it silently fails and gets disabled. The issue cannot be reproduced when that's the case.
- Load the Tasks webapp in browser (I use Chrome).
- Open the app's devtools console. Switch to the network panel, filter for 'websocket' and locate the websocket which is used to send DDP messages from the server to the client. It should contain a message such as:
```
a["{\"msg\":\"connected\",\"session\":\"2p8HQNwbA2ErPGCb4\"}"]
```
- Sign in, create a new task
- Verify that the DDP traffic contains a message for the newly created task
- Using a mongo shell (e.g. Robo3T) run a query to find the newly created task, register that Task's id.
- Issue the following update command:
```js
db.getCollection('tasks').updateOne({_id: 'newlyCreatedTaskId'}, {$set: {hash: {a: 1, b:2, c: 3}}})
```
- Confirm the DDP channel shows a similar `changed` message:
```
a["{\"msg\":\"changed\",\"collection\":\"tasks\",\"id\":\"AmiAQ9i3NkE6eKHmh\",\"fields\":{\"hash\":{\"a\":1,\"b\":2,\"c\":3}}}"]
```
- Then issue the following update:
```js
db.getCollection('tasks').updateOne({_id: 'newlyCreatedTaskId'}, {$set: {hash: {b:2, c: 3}}})
```
**Expected behavior**
A new DDP message confirming the change should immediately appear. 

**Actual behavior**
No message appears. The client never gets notified that the `a` field was removed.

### Notes:
- If `MONGO_OPLOG_URL` is misconfigured or oplog tailing is disabled - after a short delay of a few seconds, the expected update will be sent over ddp. 
- If there's a problem viewing the DDP channel, you may confirm in a different method the subscription doesn't update the client.  Add to the client code somewhere:
```js
// I added it in TasksPage.jsx
window.TasksCollection = TasksCollection;
```
Then in the devtools console, inspect the task document before/after the change via:
```js
TasksCollection.find().fetch()[0]
```

**Comment by Grubba27** [comment]
Hello there! have tested it, and this issue actually happens. Also, I've thought of a workaround for your problem:

```jsx
{Array.isArray(task.hash) && task?.hash?.map((item, index) => {
        const key = Object.keys(item).at(0);
        return (<span key={`${index}_${new Date().toISOString()}`}>{` ${key} - ${item[key]}  `}</span>);
      })}
```

as you mentioned, that may change the keys included in your hash, so I thought of using an array.

Also, if the keys must stay in the object, you can, instead of deleting them, if you set them to undefined or null, it would still work(meteor gets notified)

While this solution is just a workaround is the best that I could think of for the moment.

**Comment by igal1c0de4n** [comment]
@Grubba27 thanks for taking the time to reproduce and the confirmation.

Re your suggestion:
I thought about converting the hash into an array. It'll require a non-trivial db migration, including changing code of non-meteor (nodejs) microservices. I'd need to implement it across several mongodb collections because nested dictionaries (hash) are used in quite a few places. It could end up a few days change, with a day or two at least of stabilization/debugging period, not to mention potential risk of prd ricochets.

With regards to $set the fields to null or undefined (or use $unset): these are all valid solutions except some nodejs code writes the complete document after processing it somewhat externally (applying deep-diff deltas) hence it'll be non-trivial to achieve complete control over the nested parameter of the $set which eventually sets the value of the nested hash.

All in all, it seems I have no choice but to wait for the Meteor release which fixes it.

This brings me to the question: perhaps it is already fixed in a yet-unreleased Meteor version?
According to [this comment](https://github.com/meteor/meteor/issues/12114#issuecomment-1206156548) this may already be the case. I just can't tell if to expect it in v2.7.4 or v2.8.0, if indeed this issue is already solved.

**Comment by radekmie** [comment]
@igal1c0de4n I've checked it now, and #12107 does **not** solve the issue, but #12116 does. Surprisingly enough, the current implementation (i.e., the one that is available in the 2.8-beta.0 version) _does_ handle adding new fields (`{a: 1}` -> `{a: 1, b: 2}`), but doesn't removing them (`{a: 1, b: 2}` -> `{a: 1}`). It also works when the object becomes empty (`{a: 1}` -> `{}`).

@Grubba27 As #12116 solves this issue, it'd be good to review and release it in the next beta version.

**Comment by igal1c0de4n** [comment]
Deploy 2.8-beta.0 on Elastic Beanstalk [Node.js 14 running on 64bit Amazon Linux 2/5.5.5]
throws from `mongo_driver`:
```
Aug 10 10:52:47 ip-172-31-35-48 web: > mapme-meteor-editor-app@0.1.0 start /var/app/current
Aug 10 10:52:47 ip-172-31-35-48 web: > node main.js
Aug 10 10:52:48 ip-172-31-35-48 web: /var/app/current/programs/server/node_modules/fibers/future.js:280
Aug 10 10:52:48 ip-172-31-35-48 web: throw(ex);
Aug 10 10:52:48 ip-172-31-35-48 web: ^
Aug 10 10:52:48 ip-172-31-35-48 web: TypeError: Cannot read property 'on' of undefined
Aug 10 10:52:48 ip-172-31-35-48 web: at new MongoConnection (packages/mongo/mongo_driver.js:203:24)
Aug 10 10:52:48 ip-172-31-35-48 web: at new MongoInternals.RemoteCollectionDriver (packages/mongo/remote_collection_driver.js:4:16)
Aug 10 10:52:48 ip-172-31-35-48 web: at Object.<anonymous> (packages/mongo/remote_collection_driver.js:36:10)
Aug 10 10:52:48 ip-172-31-35-48 web: at Object.defaultRemoteCollectionDriver (packages/underscore.js:784:19)
Aug 10 10:52:48 ip-172-31-35-48 web: at new Collection (packages/mongo/collection.js:110:40)
Aug 10 10:52:48 ip-172-31-35-48 web: at new AccountsCommon (packages/accounts-base/accounts_common.js:43:18)
Aug 10 10:52:48 ip-172-31-35-48 web: at new AccountsServer (packages/accounts-base/accounts_server.js:29:5)
Aug 10 10:52:48 ip-172-31-35-48 web: at packages/accounts-base/server_main.js:7:12
Aug 10 10:52:48 ip-172-31-35-48 web: at module (packages/accounts-base/server_main.js:19:1)
Aug 10 10:52:48 ip-172-31-35-48 web: at fileEvaluate (packages/modules-runtime.js:336:7)
Aug 10 10:52:48 ip-172-31-35-48 web: at Module.require (packages/modules-runtime.js:238:14)
Aug 10 10:52:48 ip-172-31-35-48 web: at require (packages/modules-runtime.js:258:21)
Aug 10 10:52:48 ip-172-31-35-48 web: at /var/app/current/programs/server/packages/accounts-base.js:2397:15
Aug 10 10:52:48 ip-172-31-35-48 web: at /var/app/current/programs/server/packages/accounts-base.js:2404:3
Aug 10 10:52:48 ip-172-31-35-48 web: at /var/app/current/programs/server/boot.js:407:38
Aug 10 10:52:48 ip-172-31-35-48 web: at Array.forEach (<anonymous>)
Aug 10 10:52:48 ip-172-31-35-48 web: npm ERR! code ELIFECYCLE
Aug 10 10:52:48 ip-172-31-35-48 web: npm ERR! errno 1
Aug 10 10:52:48 ip-172-31-35-48 web: npm ERR! mapme-meteor-editor-app@0.1.0 start: `node main.js`
Aug 10 10:52:48 ip-172-31-35-48 web: npm ERR! Exit status 1
```
Are there plans to release the oplog fixes to v2.7.4?
It enables an oplog-fix deploy without bleeding edge changes.

**Comment by radekmie** [comment]
@igal1c0de4n This error is reported is #12118 (I already pinged @denihs about it).

**Comment by igal1c0de4n** [comment]
Hi. 
As this issue is affecting my production app my customers need me to provide an update re the estimated shipping of a fix. 
Can someone shed light on which Meteor version is expected to fix this issue, and what is it's ETA?

**Comment by radekmie** [comment]
@igal1c0de4n The latest Meteor 2.8 beta should already include this fix.

**Comment by igal1c0de4n** [comment]
I guess no v2.7.x will ship with this fix. 
Unfortunately I'll be able to test it only after https://github.com/meteor/meteor/issues/12118 is released

**Comment by radekmie** [comment]
@igal1c0de4n This should also be fixed in the 2.8 beta.

**Comment by nachocodoner** [comment]
This issue was resolved in Meteor 2.8 when proper MongoDB 5.0+ v2 oplog parsing was merged (via PR https://github.com/meteor/meteor/pull/12116 / [#12107](https://github.com/meteor/meteor/pull/12107)). Also, Meteor 3.5 has introduced change stream in favor of old oplog. We recommend to upgrade and try it out. Closing as fixed.
